### PR TITLE
Add Logistic and RCF algorthims

### LINKED ISSUE #826: [FEATURE] Add LogisticRegression algorithm to ML Commons
### Is your feature request related to a problem? 
We are adding to add a classification algorithm in OpenSearch ml-common plugin. Customers can use this algorithm to train a LogisticRegression model using their data indexed in openSearch clusters, and make predictions to help them make binary classification decisions.

We would like to get this documented with the 2.2 release.

### Do you have any additional context?
https://github.com/opensearch-project/ml-commons/issues/318#

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
Signed-off-by: Naarcha-AWS <naarcha@amazon.com>

Fixes #826 and part of https://github.com/opensearch-project/documentation-website/issues/806.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _ml-commons-plugin/algorithms.md [hunk 1]
```diff
@@ -316,4 +400,218 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+
+A classification algorithm, logistic regression models the probability of a discrete outcome given an input variable. In ML commons, these classifications include both binary and multi-class. The most common is the binary classification, which takes two values such as "true/false" or "yes/no", and predicts the outcome based on the values specified. On the other hand, a multi-class output can categorize different inputs based on type. This makes logistic regression most useful for situations where you are trying to determine how your inputs fit best into a specified-category. 
+
+### Parameters
+
+| Parameter | Type | Description | Default Value |
+|---|---|---|---|
+| learningRate | Double | The rate of speed the gradient moves during gradient descent | 1 |
+| momentumFactor | Double | The idea that a regressor that had recently risen or fallen during convergence will continue that trend over the medium term | 0 |
+| epsilon | Double | The criteria that a linear model is identified | 0.1 |
```
Comment (cwillum) [inline _ml-commons-plugin/algorithms.md:415]: Does this description mean one of these two:
"The criteria that determines whether a linear model is identified"?
"The criteria that identifies a linear model"?

File: _ml-commons-plugin/algorithms.md [hunk 2]
```diff
@@ -316,4 +400,218 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+* [Train]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/#train-model)
+* [Predict]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/#predict)
+
+### Example: Tran/Predict with IRIS Data
+
+The following example creates an index in OpenSearch with [Iris data](https://archive.ics.uci.edu/ml/datasets/iris), then trains the data using logistic regression. Lastly, it uses the trained model to predict Iris types separated by row.
+
+#### Create an Iris index
+
+Before using this request, please make that you've downloaded [Iris data](https://archive.ics.uci.edu/ml/datasets/iris).
```
Comment (cwillum) [inline _ml-commons-plugin/algorithms.md:438]: add a "sure" here > make sure

File: _ml-commons-plugin/algorithms.md [hunk 3]
```diff
@@ -316,4 +400,218 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+| epsilon | Double | The criteria that a linear model is identified | 0.1 |
+| beta1 | Double | The exponential decay rates for the moment estimates | 0.9 |
+| beta2 | Double | The exponential decay rates for the moment estimates | 0.99 |
+| decayRate | Double | The Root Mean Squared Propagation (RMSProp) | 0.9 |
+| momentumType | MomentumType | The momentum with SDG to help accelerate gradients vectors in the right directions, leading to faster convergence between vectors | STANDARD |
+| optimizerType | OptimizerType | The optimizer used in the model  | AdaGrad |
+| target | String | The target field | null |
+| objectiveType | ObjectiveType | The objective function type | LogMulticlass |
+| epochs | Integer | The number of iterations | 5 |
+| batchSize | Integer | The size of minbatches | 1 |
```
Comment (cwillum) [inline _ml-commons-plugin/algorithms.md:424]: Is this properly "minibatches"?

**Review by cwillum — APPROVED** [review]
Looks good. Just a few comments/queries.

File: _ml-commons-plugin/algorithms.md [hunk 4]
```diff
@@ -198,6 +198,90 @@ time_zone | string | The time zone for the time_field field | "UTC"
...
+
+* [Train]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/#train-model)
+* [Predict]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/#predict)
+* [Train and predict]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/#train-and-predict)
+
+### Example: Train and Predict
+
+The following example estimates cluster centers and gives cluster labels for each sample in a given data frame.
+
+```bash
```
Comment (ylwu-amzn) [inline _ml-commons-plugin/algorithms.md:222]: Seems wrong json format.  Should be
```
POST _plugins/_ml/_train_predict/RCF_SUMMARIZE
{
  "parameters": {
    "centroids": 3,
    "max_k": 15,
    "distance_type": "L2"
  },
  "input_data": {
    "column_metas": [
      {
        "name": "d0",
        "column_type": "DOUBLE"
      },
      {
        "name": "d1",
        "column_type": "DOUBLE"
      }
    ],
    "rows": [
      {
        "values": [
          {
            "column_type": "DOUBLE",
            "value": 6.2
          },
          {
            "column_type": "DOUBLE",
            "value": 3.4
          }
        ]
      }
    ]
  }
}
```

File: _ml-commons-plugin/algorithms.md [hunk 5]
```diff
@@ -316,4 +400,218 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+
+## Logistic regression
+
+A classification algorithm, logistic regression models the probability of a discrete outcome given an input variable. In ML commons, these classifications include both binary and multi-class. The most common is the binary classification, which takes two values such as "true/false" or "yes/no", and predicts the outcome based on the values specified. On the other hand, a multi-class output can categorize different inputs based on type. This makes logistic regression most useful for situations where you are trying to determine how your inputs fit best into a specified-category. 
+
+### Parameters
+
+| Parameter | Type | Description | Default Value |
+|---|---|---|---|
+| learningRate | Double | The rate of speed the gradient moves during gradient descent | 1 |
```
Comment (kolchfa-aws) [inline _ml-commons-plugin/algorithms.md:413]: Is this the "Gradient descent step size"?

File: _ml-commons-plugin/algorithms.md (hunk 1 above)
Comment (kolchfa-aws) [reply _ml-commons-plugin/algorithms.md:415]: Epsilon normally denotes the error... Is this the case here?

File: _ml-commons-plugin/algorithms.md [hunk 6]
```diff
@@ -316,4 +400,218 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+| momentumFactor | Double | The idea that a regressor that had recently risen or fallen during convergence will continue that trend over the medium term | 0 |
+| epsilon | Double | The criteria that a linear model is identified | 0.1 |
+| beta1 | Double | The exponential decay rates for the moment estimates | 0.9 |
+| beta2 | Double | The exponential decay rates for the moment estimates | 0.99 |
+| decayRate | Double | The Root Mean Squared Propagation (RMSProp) | 0.9 |
+| momentumType | MomentumType | The momentum with SDG to help accelerate gradients vectors in the right directions, leading to faster convergence between vectors | STANDARD |
```
Comment (kolchfa-aws) [inline _ml-commons-plugin/algorithms.md:419]: help accelerate **gradient** vectors

File: _ml-commons-plugin/algorithms.md (hunk 5 above)
Comment (Zhangxunmt) [reply _ml-commons-plugin/algorithms.md:413]: Yes it's the step size at each iteration while moving toward a minimum of a loss function or optimal value.

Comment (kolchfa-aws) [reply _ml-commons-plugin/algorithms.md:413]: Thank you, @Zhangxunmt. This is a better way to explain the parameter.

File: _ml-commons-plugin/algorithms.md (hunk 1 above)
Comment (Zhangxunmt) [reply _ml-commons-plugin/algorithms.md:415]: The momentum factor is a value between 0 and 1, and it is an extra weight added onto the weight factors that accelerates the rate at which the weight factors are adjusted which helps move the minimization routine out of local minima.
The epsilon is the value for stabilising the gradient inversion.

File: _ml-commons-plugin/algorithms.md (hunk 3 above)
Comment (Zhangxunmt) [reply _ml-commons-plugin/algorithms.md:424]: Yes it is minimum batches. @Naarcha-AWS I have updated the parameters descriptions in the quip doc. Can you please check and update them?

File: _ml-commons-plugin/algorithms.md (hunk 1 above)
Comment (kolchfa-aws) [reply _ml-commons-plugin/algorithms.md:415]: Thank you!

File: _ml-commons-plugin/algorithms.md (hunk 2 above)
Comment (Zhangxunmt) [inline _ml-commons-plugin/algorithms.md:438]: I have updated this section to include both creating index and ingesting data in the quip doc (https://quip-amazon.com/QqsaA3lE0Sgd/ML-Commons-Functional-Test). Can you pls update accordingly?

**Review by kolchfa-aws — APPROVED** [review]
LGTM after implementing all comments :)

File: _ml-commons-plugin/algorithms.md [hunk 7]
```diff
@@ -316,4 +403,230 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+
+| Parameter | Type | Description | Default Value |
+|---|---|---|---|
+| learningRate | Double | The gradient descent step size at each iteration moving toward a minimum of a loss function or optimal value  | 1 |
+| momentumFactor | Double | The extra weight factors that accelerate the rate at which the weight is adjusted. This helps move the minimization routine out of local minima | 0 |
+| epsilon | Double | The value for stabilizing gradient inversion | 0.1 |
+| beta1 | Double | The exponential decay rates for the moment estimates | 0.9 |
+| beta2 | Double | The exponential decay rates for the moment estimates | 0.99 |
+| decayRate | Double | The Root Mean Squared Propagation (RMSProp) | 0.9 |
+| momentumType | MomentumType | The momentum with SDG to help accelerate gradient vectors in the right direction, leading to faster convergence between vectors | STANDARD |
```
Comment (Zhangxunmt) [inline _ml-commons-plugin/algorithms.md:422]: SDG is a typo here. should be SGD.

File: _ml-commons-plugin/algorithms.md [hunk 8]
```diff
@@ -316,4 +403,230 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+{"sepal_length_in_cm":5.1,"sepal_width_in_cm":3.5,"petal_length_in_cm":1.4,"petal_width_in_cm":0.2,"class":"Iris-setosa"}
+{ "index" : { "_index" : "iris_data" } }
+{"sepal_length_in_cm":4.9,"sepal_width_in_cm":3.0,"petal_length_in_cm":1.4,"petal_width_in_cm":0.2,"class":"Iris-setosa"}
+...
+...
+```
+
+#### Train the logistic regression model
+
+This example uses a multi-class logistic regression categorization methodology. Here, the inputs of sepal length and width, and petal length and width, are used to train the model to categorize based on the `class`, as indicated by the `target` parameter.
```
Comment (natebower) [inline _ml-commons-plugin/algorithms.md:482]: In the second sentence, can we change to "Here, the inputs of sepal and petal length and width are used..."? Can we add a noun after "categorize"? "categorize data"?

File: _ml-commons-plugin/algorithms.md [hunk 9]
```diff
@@ -316,4 +403,230 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+```json
+{
+  "model_id" : "TOgsf4IByBqD7FK_FQGc",
+  "status" : "COMPLETED"
+}
+```
+
+#### Predict results
+
+Using the `model_id` of the trained Iris dataset, logistic regression will predict what class of Iris based on the input data.
```
Comment (natebower) [inline _ml-commons-plugin/algorithms.md:523]: Change to "...logistic regression will predict the Iris class based on the input data" or "...logistic regression will predict the class of the Iris based on the input data."

File: _ml-commons-plugin/algorithms.md [hunk 10]
```diff
@@ -316,4 +403,230 @@ The API responds with the sum of the contribution and base values per aggregatio
...
+        ]
+      }
+    ]
+  }
+}
+```
+
+### Limitations
+
+Convergence metrics are not built into Tribuo's trainers. Therefore, ML commons cannot indicate the convergence status through the ML commons API.
```
Comment (natebower) [inline _ml-commons-plugin/algorithms.md:632]: Capitalize both instances of "Commons".

File: _ml-commons-plugin/algorithms.md [hunk 11]
```diff
@@ -198,6 +198,93 @@ time_zone | string | The time zone for the time_field field | "UTC"
 
 For FIT RCF, you can train the model with historical data and store the trained model in your index. The model will be deserialized and predict new data points when using the Predict API. However, the model in the index will not be refreshed with new data, because the model is fixed in time.
 
+## RCFSummarize
+
+RCFSummarize is a clustering algorithm based on the Clustering Using REpresentatives (CURE) algorithm. Compared to [k-means](#k-means), which uses random iterations to cluster, RCFSummarize uses a hierarchical-based clustering technique. The algorithm starts, with a set of randomly selected centroids larger than the centroid’s ground-truth distribution. During iteration, centroid pairs too close to each other automatically merge. Therefore, the number of centroids (`max_k`) converge to a rational number of clusters that fits ground-truth, as opposed to a fixed `k` number of clusters.
```
Comment (natebower) [inline _ml-commons-plugin/algorithms.md:203]: Instead of "hierarchical-based clustering", would just "hierarchical clustering" make sense? Remove the comma in the third sentence after "starts". If "centroid's" refers to multiple centroids, it should be "centroids'" (plural possessive). I'm not seeing "ground truth" routinely hyphenated in other sources, so let's remove the hyphen in both instances here.

**Review by natebower — COMMENTED** [review]
@Naarcha-AWS Please see my comments and let me know if you have any questions. Thanks!

File: _ml-commons-plugin/algorithms.md (hunk 8 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/algorithms.md:482]: Centriods

**Review by Zhangxunmt — APPROVED** [review]
